src/json-immutable.js

```diff
@@ -44,16 +55,57 @@ function replace(key, value) {
   let result = value
 
   if (value instanceof immutable.Record) {
-    result = replaceRecord(value)
+    result = replaceRecord(value, replace)
   }
   else if (immutable.Iterable.isIterable(value)) {
-    result = replaceIterable(value)
+    result = replaceIterable(value, replace)
   }
   else if (Array.isArray(value)) {
-    result = replaceArray(value)
+    result = replaceArray(value, replace)
   }
   else if (typeof value === 'object' && value !== null) {
-    result = replacePlainObject(value)
+    result = replacePlainObject(value, replace)
+  }
+
+  debug('result:', result, '\n---')
+  return result
+}
+
+function replaceAsync(key, value) {
+  debug('key:', key)
+  debug('value:', value)
+
+  let result = value
+
+  if (!(value instanceof Promise)) {
+    if (value instanceof immutable.Record) {
+      result = new Promise((resolve) => {
+        setImmediate(() => {
+          resolve(replaceRecord(value, replaceAsync))
+        })
+      })
+    }
+    else if (immutable.Iterable.isIterable(value)) {
+      result = new Promise((resolve) => {
+        setImmediate(() => {
+          resolve(replaceIterable(value, replaceAsync))
+        })
+      })
+    }
+    else if (Array.isArray(value)) {
+      result = new Promise((resolve) => {
+        setImmediate(() => {
+          resolve(replaceArray(value, replaceAsync))
+        })
+      })
+    }
+    else if (typeof value === 'object' && value !== null) {
+      result = new Promise((resolve) => {
+        setImmediate(() => {
```

Comment:
Ano, je async, ale není dostatečně pozdržená. `setImmediate()` zaručí, že se stihne udělat ještě jiné položky na event loopu. Bez toho to funguje podobně jako next tick.